### PR TITLE
Add anchor to titles on dash

### PR DESCRIPTION
**Pull request by mournfulCoroner**

### CODE REVIEW

**Comment by robot-charts** [comment]
**i18n-check**
<details>
<summary>The following components have not been translated</summary>
<p><a href="https://weblate.gravity-ui.com/projects/datalens/charts-9197-add-title-anchor-1784/dash-title-dialog-edit-1784/ru/">dash.title-dialog.edit (ru)</a><br><a href="https://weblate.gravity-ui.com/projects/datalens/charts-9197-add-title-anchor-1784/dash-title-dialog-edit-1784/en/">dash.title-dialog.edit (en)</a></p>
</details>

Wait for the reviewers to check your changes in Weblate and try running github action again.

File: src/ui/components/DashKit/plugins/Title/Title.tsx [hunk 1]
```diff
@@ -97,6 +108,60 @@ const titlePlugin = {
             data.text,
         ]);
 
+        const showAnchor = !disableHashNavigation && !props.editMode;
+        const withInlineAnchor = showAnchor && isInlineAnchor;
+        const withAbsoluteAnchor = showAnchor && !isInlineAnchor;
+
+        const calculateAnchor = React.useCallback(() => {
+            if (showAnchor && contentRef.current && rootNodeRef.current) {
+                const contentHeight = contentRef.current.getBoundingClientRect().height || 0;
```
Comment (flops) [inline src/ui/components/DashKit/plugins/Title/Title.tsx:117]: `.current.getBoundingClientRect()` - let's save this calculations once and use their values as vars.

File: src/ui/components/DashKit/plugins/mixins.scss [hunk 2]
```diff
@@ -0,0 +1,10 @@
+@mixin with-fade-effect() {
+    content: '';
+    position: absolute;
+    top: 0;
+    bottom: 0;
+    width: 44px;
+    right: 0;
+    mask-image: linear-gradient(to right, transparent, rgba(255, 255, 255, 0.95), rgb(0, 0, 0));
+    background-color: var(--g-color-base-generic-ultralight);
```
Comment (flops) [inline src/ui/components/DashKit/plugins/mixins.scss:9]: Let's try `background: linear-gradient` ?

File: src/ui/units/dash/utils/scrollUtils.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+}: {
+    hash: string;
+    withDelay?: boolean;
+    checkUserScroll?: boolean;
+}) => {
+    setTimeout(
+        () => {
+            // if the user scrolls the page by himself, disable scrolling
+            if (!checkUserScroll || window.scrollY === 0) {
+                scrollIntoView(hash.replace('#', ''));
```
Comment (flops) [inline src/ui/units/dash/utils/scrollUtils.ts:50]: Why are we resetting hash here?

**Review by flops — CHANGES_REQUESTED** [review]

File: src/ui/components/DashKit/plugins/RendererWrapper/RendererWrapper.tsx [hunk 4]
```diff
@@ -15,24 +15,28 @@ type RendererProps = {
...
-            >
-                {children}
-            </div>
+            <React.Fragment>
+                {beforeContentNode}
+                <div
+                    ref={nodeRef || innerNodeRef}
+                    className={b('wrapper', {
+                        [type]: Boolean(type),
+                        [String(classMod)]: Boolean(classMod),
```
[the remark this answers is not in the document]
Comment (mournfulCoroner) [reply src/ui/components/DashKit/plugins/RendererWrapper/RendererWrapper.tsx:33]: if `classMod` is `undefined`, Boolean(classMod) will be false and classMod won't be applied

File: tests/data/us-e2e-data
Comment (mournfulCoroner) [inline tests/data/us-e2e-data:1]: Disable loadingOnlyVisibleCharts setting on screenshots dash (after adding `loadedItemsMap.size === this.props.tabData?.items.length` in isLoaded condition the loading of the mobile dash was broken)

File: src/ui/components/DashKit/DashKit.ts [hunk 5]
```diff
@@ -45,7 +48,7 @@ export const getConfiguredDashKit = (pluginDefaultsGetter: typeof currentDefault
 
         const plugins = wrapPlugins(
             [
-                pluginTitle,
+                getTitlePlugin(options?.disableHashNavigation),
```
Comment (flops) [inline src/ui/components/DashKit/DashKit.ts:51]: Let's use `setSettings` like in other plugins

**Review by flops — CHANGES_REQUESTED** [review]

File: src/ui/components/DashKit/plugins/Title/AnchorLink/AnchorLink.tsx [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    const hash = `#${encodeURIComponent(to)}`;
+
+    const link = {...location, hash};
+
+    if (DL.IS_MOBILE || !show) {
+        return null;
+    }
+
+    return (
+        <React.Fragment>
```
Comment (flops) [inline src/ui/components/DashKit/plugins/Title/AnchorLink/AnchorLink.tsx:32]: Fragment is redundant here

File: src/ui/components/DashKit/plugins/mixins.scss [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+    content: '';
+    position: absolute;
+    top: 0;
+    bottom: 0;
+    width: 44px;
+    right: 0;
+    background: linear-gradient(
+        to right,
+        transparent,
+        color-mix(in sRGB, var(--g-color-base-generic-ultralight) 95%, transparent),
```
Comment (flops) [inline src/ui/components/DashKit/plugins/mixins.scss:11]: `color-mix` arrived in support in 2023... and not supported in WebView do we really has no other options?

**Review by flops — CHANGES_REQUESTED** [review]

File: src/ui/components/DashKit/plugins/Title/Title.tsx [hunk 8]
```diff
@@ -97,6 +124,56 @@ const titlePlugin = {
...
+            calculateAnchor,
+            showAnchor,
+        ]);
+
+        React.useEffect(() => {
+            const debouncedCalculateAnchor = debounce(
+                calculateAnchor,
+                WIDGET_RESIZE_DEBOUNCE_TIMEOUT,
+            );
+            window.addEventListener('resize', debouncedCalculateAnchor);
```
Comment (flops) [inline src/ui/components/DashKit/plugins/Title/Title.tsx:170]: Do we need this handler if `showAnchor === false` ?

**Review by flops — CHANGES_REQUESTED** [review]

**Review by flops — APPROVED** [review]

**Pull request merged**